error TS2339: Property 'nuxtIcon' does not exist on type 'AppConfig'

Recently upgraded Nuxt Icon in my SPA Nuxt app (using Nuxt 3.3 with SSR set to false and strict typescript) and hit a bug when running `npm run generate`.

```ts
 node_modules/nuxt-icon/dist/runtime/IconCSS.vue:6:28 - error TS2339: Property 'nuxtIcon' does not exist on type 'AppConfig'.

6 const aliases = appConfig?.nuxtIcon?.aliases || {}
                             ~~~~~~~~

node_modules/nuxt-icon/dist/runtime/Icon.vue:11:28 - error TS2339: Property 'nuxtIcon' does not exist on type 'AppConfig'.

11 const aliases = appConfig?.nuxtIcon?.aliases || {}
                              ~~~~~~~~

node_modules/nuxt-icon/dist/runtime/IconCSS.vue:25:39 - error TS2339: Property 'nuxtIcon' does not exist on type 'AppConfig'.


25   if (!props.size && typeof appConfig.nuxtIcon?.size === 'boolean' && !appConfig.nuxtIcon?.size) {
                                                                                    ~~~~~~~~

node_modules/nuxt-icon/dist/runtime/Icon.vue:28:44 - error TS2339: Property 'nuxtIcon' does not exist on type 'AppConfig'.

28 const iconName = computed(() => (appConfig.nuxtIcon?.aliases || {})[props.name] || props.name)
                                              ~~~~~~~~

node_modules/nuxt-icon/dist/runtime/IconCSS.vue:28:40 - error TS2339: Property 'nuxtIcon' does not exist on type 'AppConfig'.

28   const size = props.size || appConfig.nuxtIcon?.size || '1em'
                                          ~~~~~~~~


Found 6 errors in 2 files.

Errors  Files
     4  node_modules/nuxt-icon/dist/runtime/IconCSS.vue:6
     2  node_modules/nuxt-icon/dist/runtime/Icon.vue:11
```

Created minimal reproduction and posted code on public github repo available here: https://github.com/frasza/nuxt-icon-m-repro-bug

--- a/packages/nuxt/src/core/templates.ts
+++ b/packages/nuxt/src/core/templates.ts
@@ -222,15 +222,17 @@ declare const inlineConfig = ${JSON.stringify(nuxt.options.appConfig, null, 2)}
 type ResolvedAppConfig = Defu<typeof inlineConfig, [${app.configs.map((_id: string, index: number) => `typeof cfg${index}`).join(', ')}]>
 type IsAny<T> = 0 extends 1 & T ? true : false
 
-type MergedAppConfig<Resolved extends Record<string, any>, Custom extends Record<string, any>> = {
-  [K in keyof Resolved]: K extends keyof Custom
-    ? IsAny<Custom[K]> extends true
+type MergedAppConfig<Resolved extends Record<string, unknown>, Custom extends Record<string, unknown>> = {
+  [K in keyof (Resolved & Custom)]: K extends keyof Custom
+    ? unknown extends Custom[K]
       ? Resolved[K]
-      : Custom[K] extends Record<string, any>
-        ? Resolved[K] extends Record<string, any>
-          ? MergedAppConfig<Resolved[K], Custom[K]>
-          : Exclude<Custom[K], undefined>
-        : Exclude<Custom[K], undefined>
+      : IsAny<Custom[K]> extends true
+        ? Resolved[K]
+        : Custom[K] extends Record<string, any>
+            ? Resolved[K] extends Record<string, any>
+              ? MergedAppConfig<Resolved[K], Custom[K]>
+              : Exclude<Custom[K], undefined>
+            : Exclude<Custom[K], undefined>
     : Resolved[K]
 }
 

--- a/packages/schema/src/types/config.ts
+++ b/packages/schema/src/types/config.ts
@@ -138,7 +138,9 @@ export interface RuntimeConfig extends RuntimeConfigNamespace {
 
 // -- App Config --
 
-export interface CustomAppConfig { }
+export interface CustomAppConfig {
+  [key: string]: unknown
+}
 
 export interface AppConfigInput extends CustomAppConfig {
   /** @deprecated reserved */
@@ -158,4 +160,6 @@ export interface NuxtAppConfig {
   keepalive: boolean | KeepAliveProps
 }
 
-export interface AppConfig { }
+export interface AppConfig {
+  [key: string]: unknown
+}

--- a/test/fixtures/basic/nuxt.config.ts
+++ b/test/fixtures/basic/nuxt.config.ts
@@ -145,6 +145,20 @@ export default defineNuxtConfig({
   },
   telemetry: false, // for testing telemetry types - it is auto-disabled in tests
   hooks: {
+    'schema:extend' (schemas) {
+      schemas.push({
+        appConfig: {
+          someThing: {
+            value: {
+              $default: 'default',
+              $schema: {
+                tsType: 'string | false'
+              }
+            }
+          }
+        }
+      })
+    },
     'prepare:types' ({ tsConfig }) {
       tsConfig.include = tsConfig.include!.filter(i => i !== '../../../../**/*')
     },

--- a/test/fixtures/basic/types.ts
+++ b/test/fixtures/basic/types.ts
@@ -295,6 +295,10 @@ describe('app config', () => {
         val: number
       }
       userConfig: 123 | 456
+      someThing?: {
+        value?: string | false,
+      }
+      [key: string]: unknown
     }
     expectTypeOf<AppConfig>().toEqualTypeOf<ExpectedMergedAppConfig>()
   })

